### ISSUE TITLE
Why is this output 1 elem instead of 4 elems ?

### ISSUE THREAD

**Issue by zeroexcuses** [LaurentMazare/tch-rs#74]
```
    let y = tch::Tensor::of_slice(&[2_i64, 0, 1, 1]).view(&[4]);
    let y_pred1 =
        tch::Tensor::of_slice(&[0.1_f32, 0.2, 0.9, 1.1, 0.1, 0.2, 0.2, 2.1, 0.1, 1., 2., 3.])
            .view(&[4, 3]);
    let y_pred2 =
        tch::Tensor::of_slice(&[0.8_f32, 0.2, 0.3, 0.2, 0.3, 0.5, 0.2, 0.2, 0.5, 1., 2., 3.])
            .view(&[4, 3]);

    println!(
        "loss1: {:?}, loss2: {:?}",
        y_pred1.cross_entropy_for_logits(&y),
        y_pred2.cross_entropy_for_logits(&y),
    );
```

Is giving me output:

```
loss1: [0.7243779897689819], loss2: [1.281076192855835]
```

This surprises me, as I expect output of 4 elems, but only get 1 elem output.

**Comment by zeroexcuses** [comment]
Hmm, it looks like the reason is
```
    pub fn cross_entropy_for_logits(&self, targets: &Tensor) -> Tensor {
        self.log_softmax(-1).nll_loss(&targets)
    }
```
which "flattens" the original vector. This explains why the output is 1 instead of 4 elems -- but, how do we get a 4 elem output?

**Comment by LaurentMazare** [comment]
As you noted, this uses `nll_loss` which by default takes the element-wise mean of the loss.
Rather than using `nll_loss` you could do something like the following.

```rust
self.log_softmax(-1).g_nll_loss::<Tensor>(&targets, None, Reduction::None, -100)
```

**Comment by zeroexcuses** [comment]
1. Thanks for the quick response.

2. There's an even bigger screw up on my part, I didn't write the PyTorch code first:

```

loss = torch.nn.CrossEntropyLoss()
input = torch.randn(3, 5, requires_grad=True)
target = torch.empty(3, dtype=torch.long).random_(5)
output = loss(input, target)

print(output)
```

I incorrectly thought that PyTorch was outputting a 3-elem output, whereas the PyTorch code also only outputs 1 elem.

Closing issue as resolved.
